src/buildMiddleware.ts

```diff
@@ -70,11 +69,58 @@ export function buildMiddleware<Definitions extends EndpointDefinitions, Reducer
       startNextPoll(action.meta.arg, mwApi);
     }
 
+    if (onFocus.match(action)) {
+      refetchValidQueries(mwApi, 'refetchOnFocus');
+    }
+    if (onOnline.match(action)) {
+      refetchValidQueries(mwApi, 'refetchOnReconnect');
+    }
+
     return result;
   };
 
   return { middleware };
 
+  function refetchQuery(
+    querySubState: Exclude<QuerySubState<any>, { status: QueryStatus.uninitialized }>,
+    queryCacheKey: string,
+    override: Partial<QueryThunkArg<any>> = {}
+  ) {
+    return queryThunk({
+      endpoint: querySubState.endpoint,
+      originalArgs: querySubState.originalArgs,
+      internalQueryArgs: querySubState.internalQueryArgs,
+      subscribe: false,
+      forceRefetch: true,
+      startedTimeStamp: Date.now(),
+      queryCacheKey: queryCacheKey as any,
+      ...override,
+    });
+  }
+
+  function refetchValidQueries(api: MWApi, type: 'refetchOnFocus' | 'refetchOnReconnect') {
+    const state = api.getState()[reducerPath];
+    const queries = state.queries;
+    const subscriptions = state.subscriptions;
+
+    batch(() => {
+      for (const queryCacheKey of Object.keys(subscriptions)) {
+        const querySubState = queries[queryCacheKey];
+        const subscriptionSubState = subscriptions[queryCacheKey];
+
+        if (!subscriptionSubState || !querySubState || querySubState.status === QueryStatus.uninitialized) return;
+
+        const shouldRefetch =
+          Object.values(subscriptionSubState).some((sub) => sub[type] === true) ||
+          (Object.values(subscriptionSubState).every((sub) => sub[type] === undefined) && state.config[type]);
+
+        if (shouldRefetch) {
```

Comment:
This would currently also refetch unmounted queries. Should we do that or remove those instead?

Reply:
No, that was wrong. But still a question here:

This only loops over subscribed queries, which means that it will only refetch actually mounted stuff and stuff that is not mounted is neither removed nor refreshed, potentially leading to semi-stale stuff.
So, we could either refresh those or remove them, like we do in `invalidateEntities`. Thoughts? :thinking:

Reply:
I think we should probably refresh them.... but I could also just see the better fit for dropping them? Tough call.

Reply:
In the spirit of https://react-query.tanstack.com/docs/guides/scroll-restoration we should maybe refetch them by default, but add a config option to remove instead?
So a `refetchUnmounted` option that kicks in both here and on invalidateEntities?